[mlir][Transforms] Add missing check in tosa::transpose::verify()

DarshanRamakant opened the pull request:
The tosa::transpose::verify() should make sure
that the permutation numbers are within the size of 
the input array. Otherwise it will cause a cryptic array
out of bound assertion later.Fix #99513.

darshan-opensource committed:
[mlir][Transforms] Add missing check in tosa::transpose::verify()

The tosa::transpose::verify() should make sure
that the permutation numbers are within the size
of the input array. Otherwise it will cause a
cryptic array out of bound assertion later.

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:
```diff
@@ -0,0 +1,10 @@
+	%arg0 = tensor.empty() : tensor<3x4x5xi32>
+	%1110 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
+	%143 = tosa.transpose %arg0, %1110: (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
+	return 
```
nit: Could you please format your test as well?

DarshanRamakant replied:
I will modify the testcase

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:
```diff
@@ -0,0 +1,10 @@
+func.func @func1() {
+	%arg0 = tensor.empty() : tensor<3x4x5xi32>
+	%1110 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
+	%143 = tosa.transpose %arg0, %1110: (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
```
It seems strange to me: why tosa.transpose verify passed?

DarshanRamakant replied:
Yes you are right, there was no check in the tosa.transpose verifier. I have added a check there as well now.

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:
```diff
@@ -0,0 +1,10 @@
+// RUN: not --crash mlir-opt %s --pass-pipeline="builtin.module(func.func(tosa-to-linalg-named))" 2>&1 | FileCheck -check-prefix=CHECK %s
```
I don't think we should put the test in this file; util and this conversion are not closely related.

DarshanRamakant replied:
Removed this test, added this case in an existing test about invalid cases.

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalidperm.mlir:
```diff
@@ -0,0 +1,10 @@
+// RUN: not --crash mlir-opt %s --pass-pipeline="builtin.module(func.func(tosa-to-linalg-named))" 2>&1 | FileCheck -check-prefix=CHECK %s
+
+func.func @func1() {
+	%arg0 = tensor.empty() : tensor<3x4x5xi32>
```
You've used very casual names for your functions and variables. This is not recommended.

DarshanRamakant replied:
I will modify.

cxy-1993 commented on mlir/include/mlir/Dialect/Utils/IndexingUtils.h:
```diff
@@ -202,6 +202,9 @@ SmallVector<T> applyPermutation(ArrayRef<T> input,
          "expected input rank to equal permutation rank");
+  assert(
+      llvm::all_of(permutation, [&](size_t s) { return s < input.size(); }) &&
+      "permutation must be within input bounds");
```
Top level question: Do we really need to do this check? The permutation itself represents its value range. If there's an error, it should have been caught before the call site. In what scenarios would we need this check?

DarshanRamakant replied:
I feel this check is needed. If you look at the original issue ([https://github.com/llvm/llvm-project/issues/99513]) there was a crash with cryptic assertion fail. This is happening because of the way permutation range is constructed :
`  auto permutationRange = llvm::map_range(
      llvm::seq<unsigned>(0, input.size()),
      [&](int64_t idx) -> T { return input[permutation[idx]]; });`

We want to make sure that we are not accessing out of boud while doing "input[permutation[idx]]"

cxy-1993 reviewed (changes requested):
Thank you for your contribution. Personally, I don't think this assert is necessary. Perhaps we need to check the verify of other dialects.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Utils/IndexingUtils.h:
```diff
@@ -202,6 +202,9 @@ SmallVector<T> applyPermutation(ArrayRef<T> input,
                                 ArrayRef<int64_t> permutation) {
   assert(input.size() == permutation.size() &&
          "expected input rank to equal permutation rank");
+  assert(
```
Would you be able to change the result of this to be a `FailureOr<SmallVector<T>>` That way, we can avoid the asserts

DarshanRamakant replied:
That change can be done. But that would lead to :
1) Changing the code of all the callers of "applyPermutation()" to take care of the failue path. 
2) Each caller can throw different error message on failure, which can lead to bad user experience. 

The original aim of this patch was to throw a meaningful error message. 
I can change the function signature in a new patch if all the reviewers agrees for that.

DarshanRamakant commented:
@cxy-1993 @MaheshRavishankar  can you please merge this change if there are no other review comments ?

DarshanRamakant commented:
Mistakenly closed the pull request, reopening it again.

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir:
```diff
@@ -36,3 +36,14 @@ func.func @rfft2d_with_non_float_type(%arg0 : tensor<1x1x1xi32>) -> (tensor<1x1x
+  // expected-error@+3 {{permutation must be within input bounds}}
+	%14 = tensor.empty() : tensor<3x4x5xi32>
+	%c1 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
+	%72 = tosa.transpose %14, %c1 : (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
```
This test should be placed in mlir/test/Dialect/Tosa/invalid.mlir

DarshanRamakant replied:
Updated

cxy-1993 commented on mlir/test/Conversion/TosaToLinalg/tosa-to-linalg-invalid.mlir:
```diff
@@ -36,3 +36,14 @@ func.func @rfft2d_with_non_float_type(%arg0 : tensor<1x1x1xi32>) -> (tensor<1x1x
+  // expected-error@+3 {{permutation must be within input bounds}}
+	%14 = tensor.empty() : tensor<3x4x5xi32>
+	%c1 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
+	%72 = tosa.transpose %14, %c1 : (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
```
You should pay more attention to the test format: reduce the use of tabs. Additionally, the variable names are too casual. %c1 doesn't match your value meaning at all, it is confusing. If you're unsure about naming variables, it's recommended to start with %0.

DarshanRamakant replied:
Thanks for the feedback. I have used the test case as it is from the original issue. I will modify the names to use it from %0. 
Having said that many of the existing test cases ( even in the same folder) are not following naming convention that you have recommended.

cxy-1993 commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -1116,6 +1116,12 @@ LogicalResult tosa::TransposeOp::verify() {
     if (!isPermutationVector(constantPerms))
       return emitOpError() << "expected valid permutation tensor";
+
+    if (inputType.hasRank() && (!inputType.getNumDynamicDims()) &&
```
why no dynamic dims?

DarshanRamakant replied:
Without that the existing behaviour will change. Couple of test cases in the shape-inference will give a different behaviour without that check
`https://github.com/llvm/llvm-project/blob/50a2b31800a554b35367a5f3bcebb640703a48e1/mlir/test/Dialect/Tosa/tosa-infer-shapes.mlir#L1398`
Above testcase will be having a different behaviour if we do it for dynamic shapes. So I tried to retain the original behaviour with that check.

DarshanRamakant commented:
@cxy-1993  @MaheshRavishankar @Mogball can you please merge this PR ?

cxy-1993 commented:
I maintain my original stance: when permutation is constant, verify doesn't require static shape. So I cant merge this for you. As for the test case you highlighted, I believe it would be more appropriate to modify it, such as by introducing perm as a parameter. And I don't think that's a good test, neither in terms of its style nor its test point design.

DarshanRamakant commented:
> I maintain my original stance: when permutation is constant, verify doesn't require static shape. So I cant merge this for you. As for the test case you highlighted, I believe it would be more appropriate to modify it, such as by introducing perm as a parameter. And I don't think that's a good test, neither in terms of its style nor its test point design.

Aa suggested, I have removed the dynamic shape check, and modified the failing tests.

cxy-1993 reviewed (approved):
Thanks for iterating on this patch, LGTM.

DarshanRamakant commented:
> Thanks for iterating on this patch, LGTM.

Thanks @cxy-1993, can you please merge the PR ?

cxy-1993 commented:
> > Thanks for iterating on this patch, LGTM.
> 
> Thanks @cxy-1993, can you please merge the PR ?

Yes. Could you please revise the commit message and PR description before merging? The information in them no longer aligns with the current code.

DarshanRamakant commented:
> > > Thanks for iterating on this patch, LGTM.
> > 
> > 
> > Thanks @cxy-1993, can you please merge the PR ?
> 
> Yes. Could you please revise the commit message and PR description before merging? The information in them no longer aligns with the current code.

Done

cxy-1993 commented on mlir/test/Dialect/Tosa/invalid.mlir:
```diff
@@ -413,3 +413,38 @@ func.func @test_tile_invalid_multiples() {
+	%0 = tensor.empty() : tensor<3x4x5xi32>
+	%1 = arith.constant dense<[3, 0, 1]> : tensor<3xi32>
+	%2 = tosa.transpose %0, %1 : (tensor<3x4x5xi32>, tensor<3xi32>) -> tensor<3x4x5xi32>
+	return 
```
Do not use tab

DarshanRamakant replied:
Strangely it was no visible in my local vs code. Fixed the issue now.
Yes, the latest CI failures not related, it is happening on all the patches.

cxy-1993 reviewed (commented):
I've found another style issue... There's a problem with the CI in the LLVM, this patch isn't the cause of the current CI failure. Please fix this style issue, and then I'll help you merge the code.

darshan-opensource committed:
Merge branch 'main' into main
